File: MapEditor/MapEditor/MapParser.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace MapEditor
{
	public class MapParser
	{
		private string filename;
		Dictionary<string, TileTemplate> tileLookup;
		public MapParser(string filename, Dictionary<string, TileTemplate> tileLookup)
		{
			this.filename = filename;
			this.tileLookup = tileLookup;
		}

		public Model Parse()
		{
			Dictionary<string, string> values = new Dictionary<string, string>();
			string[] lines = System.IO.File.ReadAllText(this.filename).Replace("\r\n", "\n").Replace('\r', '\n').Split('\n');
			foreach (string line in lines)
			{
				string[] parts = line.Split(':');
				if (parts.Length >= 2)
				{
					string key = parts[0];
					if (key.Length > 0 && key[0] == '#')
					{
						key = key.Substring(1);
						string value = parts[1];
						for (int i = 2; i < parts.Length; ++i)
						{
							value += ':' + parts[i];
						}

						values[key] = value;
					}
				}
			}

			int width = int.Parse(values["width"]);
			int height = int.Parse(values["height"]); // let it crash if this is wrong.

			Model model = new Model(width, height);

			TileTemplate[] topLayer = new TileTemplate[width * height];
			TileTemplate[] bottomLayer = new TileTemplate[width * height];

			string[] tileIdsUpper = values["upper"].Split(',');
			string[] tileIdsLower = values["lower"].Split(',');

			for (int i = width * height - 1; i >= 0; --i)
			{
				string tileId = tileIdsUpper[i].Trim();
				if (tileId.Length > 0)
				{
					topLayer[i] = this.tileLookup[tileId];
				}

				tileId = tileIdsLower[i].Trim();
				if (tileId.Length > 0)
				{
					bottomLayer[i] = this.tileLookup[tileId];
				}
			}

			// TODO: other fields

			model.SetTiles(topLayer, true);
			model.SetTiles(bottomLayer, false);

			model.SetRawFileData(values);

			model.Path = filename;
			model.IsSideScroll = !values.ContainsKey("view") || values["view"].Trim().ToLowerInvariant() != "over";

			return model;
		}
	}
}

File: trunk/source/MapParser.py
class MapParser:
	def __init__(self, file):
		path = ('maps/' + file).replace('/', os.sep).replace('\\', os.sep)
		c = open(path, 'rt')
		t = c.read()
		c.close()
		values = {}
		for line in t.split('\n'):
			parts = line.split(':')
			if len(parts) < 2:
				continue
			
			key = trim(parts[0][1:])
			value = trim(':'.join(parts[1:]))
			values[key] = trim(value)
		self.values = values
	
	def parse(self):
		m = StemCell()
		m.width = int(self.values['width'])
		m.height = int(self.values['height'])
		m.side = self.values['view'].lower() == 'side'
		m.upper = self.getTiles(self.values['upper'], m.width, m.height)
		m.lower = self.getTiles(self.values['lower'], m.width, m.height)
		return m
		
		
	def getTiles(self, ids, width, height):
		output = []
		for id in map(trim, ids.split(',')):
			if id == '':
				output.append(None)
			else:
				output.append(getTile(id))
		
		grid = makeGrid(width, height)
		y = 0
		index = 0
		while y < height:
			x = 0
			while x < width:
				grid[x][y] = output[index]
				index += 1
				x += 1
			y += 1
		return grid